### ISSUE TITLE
bundle config exits with successful exit status even when requested setting isn't set

### ISSUE THREAD

**Issue by jasonkarns** [rubygems/rubygems#3215]
`bundle config SOME_SETTING` is expected to print the value for the requested setting.

If the setting has a value, it is printed to STDOUT and exit status of 0 is returned as expected. If the requested setting is _not_ set or has no value, it is expected that a non-zero exit status is returned. (as is conventional for cli programs) However, an exit status of 0 is returned in all cases.

This means that scripts and utilities that wish to interrogate bundler to find the value for a given setting have to parse the STDOUT of `bundle config` in order to determine if a particular setting is unset.

```sh
$ bundle config something
Settings for `something` in order of priority. The top value will be used
You have not configured a value for `something`
$ echo $?
0
```

Further, if any work is done on this, I would strongly request that the output from this command go to STDERR.

I would expect `bundler config foo 2>/dev/null` to print nothing but the final computed value of `foo`. All the rest of bundler's output (which is indeed super helpful), should go to STDERR as it is extraneous to the core role of `bundle-config`. And thus `bundler config UNSET 2>/dev/null` would print no output and exit non-zero.

**Comment by jeromedalbert** [comment]
The last comment was 6 years ago, should this issue be closed?

**Comment by deivid-rodriguez** [comment]
In my opinion, this request still makes sense and makes `bundle config` more useful, even if it's old.

**Comment by jeromedalbert** [comment]
Ok! There are two requests here:

1. Add a non-zero exit status if the setting is not set. Samuel had some reservations about this.
2. Output informational message to stderr. There were no reservations about this and it is similar to work done in https://github.com/rubygems/rubygems/pull/7957 and https://github.com/rubygems/rubygems/pull/7966 to emit progress to stderr. If this request makes sense I can make a PR.

**Comment by deivid-rodriguez** [comment]
Yep, a PR would be great!

I think the fact that Bundler can't distinguish between a explicit `nil` setting and an "unset" setting doesn't contradict this proposal?

There are of course backwards compatibility concerns when making a command that previously succeeded start failing, but I think it will probably affect few people, an maybe actually help catching typo bugs. If we do this, I'd ship it with December's minor, tagging it as a minor breaking change.

**Comment by jeromedalbert** [comment]
Regarding request 2, on second thought I think outputing the informational message to stderr is not needed here. If you want the parse the output, `bundle config something --parseable` should be used. And this command does not display any informational message so it can be easily parsed.

So that leaves us with request 1 for this issue: adding a non-zero exit status if the setting is not set.

**Comment by jasonkarns** [comment]
>  If you don't need to parse the output, it does not really matter that all output is emitted to stdout.

I think I would disagree with this assertion. As authors of the cli, we cannot foresee all the possible needs and use cases of the tool. This is why adherence to the standards and conventions (ala, a "specification") is so vital.

Example: It's exceedingly common for tools (scripts, wrappers, subshells, even just blocks of commands, etc) to silence stdout so that _they_ may be compliant even if the wrapped tools are not. In such a case, the STDOUT output may not be necessary (thus not parsed or using --parseable), but STDERR would still be expected to contain the diagnostic messages of any failures. If bundler doesn't distinguish the two, then its informational display would be swallowed unknowingly.

**Comment by jeromedalbert** [comment]
I was talking in the context of this informational message:

```
$ bundle config something
Settings for `something` in order of priority. The top value will be used
You have not configured a value for `something`
```

> STDERR would still be expected to contain the diagnostic messages of any failures.

The first line is a green informational message.
The second line displays `You have not configured a value for something` when there is no value, or `Set for the current user (/Users/jerome/.bundle/config): true` when there is a value.

Are you implying that `You have not configured a value for something` should be emitted to stderr?

**Comment by deivid-rodriguez** [comment]
What @jeromedalbert said makes sense to me, in the particular context of `bundle config`. Let's implement just request 1.

**Comment by jasonkarns** [comment]
@jeromedalbert 

> Are you implying that You have not configured a value for something should be emitted to stderr?

Yep, definitely. Otherwise, scripts and blocks that call `bundle-config` may have no indication there is an issue.

> (which doesn't seem right since the first one is not failure-related)

Strictly speaking, STDERR is supposed to contain _diagnostic_ output. Anything that is not directly related to the intent of the command.

In my opinion, `bundle-config` should be emitting the actual settings on STDOUT and the "header" message ("Settings for `something` in order of priority. The top value will be used") should be on STDERR since it's diagnostic and not part of the core responsibility of the bundle-config command. But I'd be splitting hairs with that one.

The important pieces to this request: invoking bundle-config with an unknown setting should exit with a non-zero exit code, and the message `You have not configured a value for <X>` should go to STDERR. That would make bundle-config more compliant with unix standards and subsequently easier to work with.

**Comment by dgollahon-plaid** [comment]
This bit me and a colleague today and we actually lost a fair amount of time hunting down what went wrong.

I don't think this has been mentioned so far but I was surprised to learn that `bundler config` and `bundle config set` have different semantics here: 

```
❯ bundle config foo $bar || echo 'i failed!'
Settings for `foo` in order of priority. The top value will be used
You have not configured a value for `foo`
```

v.s.

```
❯ bundle config set foo $bar || echo 'i failed!'
ERROR: "bundle config set" was called with arguments ["foo"]
Usage: "bundle config set NAME VALUE"
i failed!
```

edit: i guess this is downstream of `bundle config name` _returning_ a value and `bundle config name value` _setting_ a value. [Old versions](https://bundler.io/v1.12/man/bundle-config.1.html) of the docs exclusively use `bundle config name value`. [Newer docs](https://bundler.io/v2.7/man/bundle-config.1.html) use `bundle config set` which I assume was added for similar reasons.

Kind of feels like the non `get`/`set` subcommands should be deprecated/emit a warning.
